[REFACTOR] Example with the new version

leon3s opened the pull request:
Hey there,

I noticed some of examples wasn't working for the new version so i wanted to help by correcting them.

But i couldn't figure out how to fix the `middleware` example, the `Transform` trait seems not available anymore and im not sure what is the replacement.
As well as the `websocket` examples the `on_shutdown` callback seems not existing and it make the heartbeat stop because the variable drop if we just remote it.

Thank for your work and time.

fafhrd91 commented:
Transform is renamed to Middleware.

fafhrd91 commented:
regarding "on_shutdown". I changed FnService in new version of ntex-service, now it does not provide shutdown handler. But it is possible to create special service, something like FnShotdown, that would provide same functionality

leon3s commented:
Hey based on your commentary i update the websocket example to use the Service Trait.
What do you think of the implementation ?
If you like it i can update websocket-chat and websocket-tcp-chat using the same way

fafhrd91 commented:
I think we can move FnShutdown type to ntex-service crate. I've been thinking about something like `FnShutdown<S, F: FnOnce>`

leon3s commented:
Sure as you wish, i'll try to do it and make a pull request during the day!

leon3s commented:
So i updated example to your latest version 6.0.2 that include `fn_shutdown` and refactored the websocket example to use it.
I also corrected the middleware example but i coudn't make compile the file `read_response_body.rs` kind of lifetime problem that i coudn't figure out how to solve it..

fafhrd91 commented:
here is code:

```rust

impl<S, Err> Service<WebRequest<Err>> for LoggingMiddleware<S>
where
    S: Service<WebRequest<Err>, Response = WebResponse, Error = Error>,
    Err: 'static,
{
    type Response = WebResponse;
    type Error = Error;
    type Future<'f> = BoxFuture<'f, Result<WebResponse, Error>> where Self: 'f;

    ntex::forward_poll_ready!(service);

    fn call(&self, req: WebRequest<Err>) -> Self::Future<'_> {
        Box::pin(async move {
            self.service.call(req).await.map(|res| {
                res.map_body(move |_, body| {
                    Body::from_message(BodyLogger {
                        body,
                        body_accum: BytesMut::new(),
                    })
                    .into()
                })
            })
        })
    }
}
```

leon3s commented:
Thx i updated the middleware example.
I used :
```rust
Err: ErrorRender
```
instead of
```rust
Err: 'static
```
Maybe you don't want that ?

estin commented:
@leon3s can you look at websoket-tcp-chat example error?

WS client drop connection immediately.

on this PR
```bash
$ git log -n 1
commit 2354c61ebf630e05903fea4c50e90e6e1e47fa67 (HEAD -> fix-websocket-and-server-sent-events, origin/fix-websocket-and-server-sent-events)
Author: leon3s <leo.vernisse@gmail.com>
Date:   Wed Jan 25 13:53:29 2023 +0100

    [REFACTOR] Rustfmt
```

Server output
```bash
RUST_LOG=trace cargo run --bin websocket-tcp-chat-server
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `/home/user/tmp/fork-examples/target/debug/websocket-tcp-chat-server`
Started chat server
[2023-02-06T07:21:42Z INFO  ntex::server::builder] Starting 12 workers
[2023-02-06T07:21:42Z INFO  ntex::server::builder] Starting "tcp" service on 127.0.0.1:12345
[2023-02-06T07:21:42Z INFO  ntex::server::builder] Starting "websockets" service on 127.0.0.1:8080
[2023-02-06T07:21:42Z INFO  ntex::server::accept] Starting socket listener on 127.0.0.1:12345
[2023-02-06T07:21:42Z INFO  ntex::server::accept] Starting socket listener on 127.0.0.1:8080
Someone joined
```

Client output
```bash
RUST_LOG=trace cargo run --bin websocket-tcp-chat-client
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `/home/user/job/tmp/fork-examples/target/debug/websocket-tcp-chat-client`
[2023-02-06T07:21:48Z TRACE ntex::ws::client] Open ws connection to http://127.0.0.1:8080/ws/ addr: None
[2023-02-06T07:21:48Z TRACE ntex_connect::service] TCP connector - connecting to "127.0.0.1" port:8080
[2023-02-06T07:21:48Z TRACE ntex_connect::service] TCP connector - successfully connected to connecting to "127.0.0.1" - Some(127.0.0.1:8080)
[2023-02-06T07:21:48Z TRACE ntex::ws::client] Sending ws handshake http message
[2023-02-06T07:21:48Z TRACE ntex::ws::client] Waiting for ws handshake response
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Disconnected(None)', websocket-tcp-chat/src/client-ws.rs:21:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Panic in Arbiter thread.
```

leon3s commented:
Hey @estin I mostly tested the websocket example.

The problem seems to came from how the ID is retrieve, the unbounded channel is already disconnected when to send the ID in this function: 
`websocket-tcp-chat/src/server.rs:108`
```rust
rt::spawn(async move {
    println!("Send id back: {id}");
    let _res = sender.send(ClientMessage::Id(id)).await;
    println!("{_res:#?}");
});
```

Im not sure why i tried couple thing to fix it but i don't really have time right now i'll investigate further later.
If you remove the function that send the connect event and wait the id it work fine.
`websocket-tcp-chat/src/web.rs:70`

estin commented:
@leon3s thanks!

This is result of my research about client disconnect over http upgrade request
 - bug related to `h1/Dispatcher` and upgrade request logic
 - if comment line on commit https://github.com/ntex-rs/ntex/commit/7da66512ee8a55b4cb1c7818e3b027c610662e29#diff-5a96ae4150630a4e13848b56c1cf5f8c6660d254766b91e5ddb050bc2c47aeffR220  all works fine
 
Dispatcher stops when io has flags `IO_STOPPED | IO_STOPPING | IO_STOPPING_FILTERS | DSP_STOP`

On upgrade request called [self.io.take()](https://github.com/ntex-rs/ntex/blob/master/ntex/src/http/h1/dispatcher.rs#L509) ->  [io.take()](https://github.com/ntex-rs/ntex/blob/a1690d7b343b96ea2feb8f4b50d5563a8dfad59b/ntex-io/src/io.rs#L216)  create new io with this flags `IO_STOPPED | IO_STOPPING | IO_STOPPING_FILTERS | DSP_STOP`

There two ways to make this example works
 - don't wait client id from unbound channel https://github.com/ntex-rs/examples/pull/13#issuecomment-1418881511
 - comment line [*this.st = State::Stop;](https://github.com/ntex-rs/ntex/blob/a1690d7b343b96ea2feb8f4b50d5563a8dfad59b/ntex/src/http/h1/dispatcher.rs#L224)

Don't understand how they related.

fafhrd91 commented:
I will check later this week

fafhrd91 reviewed (approved):
thanks!

leon3s commented on form/src/main.rs:
```diff
@@ -143,11 +143,11 @@ mod tests {
     async fn handle_post_2_unit_test() {
         let state = TestRequest::default()
-            .data(AppState {
+            .state(AppState {
```
But i cannot create a new State type can i ? You removed the new method

fafhrd91 replied:
no need for State here, but to get T you have to use State<T>

leon3s replied:
Sorry im confuse this is the code and indeed your right it's not working
```rust
        let state = TestRequest::default()
            .state(AppState {
                foo: "bar".to_string(),
            })
            .to_http_request();
        let data = state.app_state::<web::types::State<AppState>>().unwrap();
```
This return `None` and crash

```rust
state.app_state::<web::types::State<AppState>>().unwrap();
```

This work, but i cannot pass it to the service for test

```rust
state.app_state::<AppState>().unwrap();
```

fafhrd91 replied:
you need State only as extractor

leon3s commented on middleware/src/redirect.rs:
```diff
@@ -22,16 +20,15 @@ pub struct CheckLoginMiddleware<S> {
     type Response = WebResponse;
     type Error = Error;
-    type Future = Either<S::Future, Ready<Result<Self::Response, Self::Error>>>;
+    type Future<'f> = Either<S::Future<'f>, Ready<Result<Self::Response, Self::Error>>> where Self: 'f;
```
I also used the `Either` from your `ntex::util`
